### PR TITLE
Eng 16984 crash during update table

### PR DESCRIPTION
**Pull request by kaikaizi**
**This does not fix ENG-16984. This is only for code cleaning and refactory.**
EE code clean up related to StringRef and ThreadLocalPool, and a few other files.
Notably replaces `pthread_key_t` with `thread_local` variables to clean clutterings, and removes unnecessary heap operations.

### CODE REVIEW

File: src/ee/common/StringRef.cpp [hunk 1]
```diff
@@ -47,8 +49,14 @@ const char* StringRef::getObject(int32_t& lengthOut) const {
                              asSizedObject(m_stringPtr)->m_size)
               << std::endl;
     // */
-    lengthOut = asSizedObject(m_stringPtr)->m_size;
-    return asSizedObject(m_stringPtr)->m_data;
+    auto const* sized = asSizedObject(m_stringPtr);
+    lengthOut = sized->m_size;
+    if (lengthOut > 0) {
+       return sized->m_data;
```
Comment (wenxuanqiu) [inline src/ee/common/StringRef.cpp]: missing a space on this line

File: src/ee/common/ThreadLocalPool.cpp [hunk 2]
```diff
@@ -96,120 +77,109 @@ ThreadLocalPool::ThreadLocalPool()
...
-void ThreadLocalPool::assignThreadLocals(const PoolLocals& mapping)
-{
+void ThreadLocalPool::assignThreadLocals(const PoolLocals& mapping) {
     vassert(mapping.enginePartitionId != NULL && getThreadPartitionId() != 16383);
 
-    pthread_setspecific(m_allocatedKey, static_cast<const void *>(mapping.allocated));
-    pthread_setspecific(m_key, static_cast<const void *>(mapping.poolData));
-    pthread_setspecific(m_stringKey, static_cast<const void*>(mapping.stringData));
-    pthread_setspecific(m_enginePartitionIdKey, static_cast<const void*>(mapping.enginePartitionId));
+    m_allocatedKey = *mapping.allocated;
```
Comment (wenxuanqiu) [inline src/ee/common/ThreadLocalPool.cpp]: I think this should be `m_allocatedKey = mapping.allocated;`

File: src/ee/common/ThreadLocalPool.cpp [hunk 3]
```diff
@@ -96,120 +77,109 @@ ThreadLocalPool::ThreadLocalPool()
...
 void ThreadLocalPool::resetStateForTest() {
-    pthread_setspecific(m_allocatedKey, NULL);
-    pthread_setspecific(m_key, NULL);
-    pthread_setspecific(m_stringKey, NULL);
-    pthread_setspecific(m_enginePartitionIdKey, NULL);
-    pthread_setspecific(m_threadPartitionIdKey, NULL);
+    m_key = nullptr;
+    m_stringKey = nullptr;
+    m_enginePartitionIdKey = nullptr;
+    m_threadPartitionIdKey = nullptr;
```
Comment (wenxuanqiu) [inline src/ee/common/ThreadLocalPool.cpp]: Is `m_allocatedKey = nullptr` intentionally left out?

File: src/ee/common/ThreadLocalPool.cpp [hunk 4]
```diff
@@ -573,8 +510,8 @@ void ThreadLocalPool::setPartitionIds(int32_t partitionId) {
 #ifdef VOLT_POOL_CHECKING
     // Don't track allocations on the mp thread because it is not used at all
     if (partitionId != 16383) {
-        pthread_mutex_lock(&s_sharedMemoryMutex);
-        PartitionBucketMap_t::iterator it = s_allocations.find(partitionId);
+        std::lock_guard<std::mutex> guard(s_sharedMemoryMutex);
+        auto const it = s_allocations.find(partitionId);
         if (it != s_allocations.end()) {
```
Comment (wenxuanqiu) [inline src/ee/common/ThreadLocalPool.cpp]: Might want to change to `if (it != s_allocations.cend()) {` to be consistent as previous changes

File: src/ee/common/StringRef.cpp [hunk 5]
```diff
@@ -85,25 +91,22 @@ int32_t StringRef::getAllocatedSizeInTempStorage() const {
 // The alternative would be to require the allocator to be aware of the
 // StringRef class so that it can call a proper accessor method.
 // An earlier implementation taking this approach proved hard to follow.
-inline StringRef::StringRef(int32_t sz)
-  : m_stringPtr(reinterpret_cast<char*>(ThreadLocalPool::allocateRelocatable(&m_stringPtr, sz)))
-{ }
+inline StringRef::StringRef(int32_t sz):
+    m_stringPtr(reinterpret_cast<char*>(ThreadLocalPool::allocateRelocatable(&m_stringPtr, sz))) {}
```
Comment (wenxuanqiu) [inline src/ee/common/StringRef.cpp]: Might want `{}` to `{ }` to be consistent with style present in this PR

File: src/ee/structures/CompactingPool.cpp [hunk 6]
```diff
@@ -32,15 +28,15 @@ CompactingPool::~CompactingPool() {
                 ThreadLocalPool::getThreadPartitionId());
         VOLT_ERROR_STACK();
 #ifdef VOLT_TRACE_ALLOCATIONS
-    BOOST_FOREACH (AllocTraceMap_t::value_type& entry, m_allocations) {
-        VOLT_ERROR("Missing deallocation for %p at:", entry.first);
-        entry.second->printLocalTrace();
-        delete entry.second;
-    }
+        for (auto const& entry, m_allocations) {
```
Comment (wenxuanqiu) [inline src/ee/structures/CompactingPool.cpp]: `for (auto const& entry : m_allocations) {`

File: src/ee/structures/CompactingPool.h [hunk 7]
```diff
@@ -112,19 +108,21 @@ namespace voltdb
...
+    static int32_t FIXED_OVERHEAD_PER_ENTRY() {
+        return static_cast<int32_t>(sizeof(Relocatable));
+    }
 
     private:
         ContiguousAllocator m_allocator;
 #ifdef VOLT_POOL_CHECKING
 #ifdef VOLT_TRACE_ALLOCATIONS
-        typedef std::unordered_map<void *, StackTrace*> AllocTraceMap_t;
+        typedef std::unordered_map<void*, StackTrace*> AllocTraceMap_t;
```
Comment (wenxuanqiu) [inline src/ee/structures/CompactingPool.h]: `using AllocTraceMap_t = std::unordered_map<void*, StackTrace*>;`

File: src/ee/structures/CompactingPool.h [hunk 8]
```diff
@@ -112,19 +108,21 @@ namespace voltdb
...
 #else
-        typedef std::unordered_set<void *> AllocTraceMap_t;
+        typedef std::unordered_set<void*> AllocTraceMap_t;
```
Comment (wenxuanqiu) [inline src/ee/structures/CompactingPool.h]: same as above

File: src/ee/common/ThreadLocalPool.cpp (hunk 2 above)
Comment (kaikaizi) [reply src/ee/common/ThreadLocalPool.cpp]: The struct `PoolLocals` has member `allocated` of type `size_t*` rather than `size_t`. Originally, this thread_local variable was also a pointer (and therefore lots of heap operations around it); I changed it from `size_t*` to `size_t`.
But, I think it's better to change things on struct `PoolLocals` as well.

File: src/ee/structures/CompactingPool.cpp (hunk 6 above)
Comment (kaikaizi) [reply src/ee/structures/CompactingPool.cpp]: Nice catch! I wonder why it compiled..

File: src/ee/common/ThreadLocalPool.cpp [hunk 9]
```diff
@@ -36,15 +36,14 @@ namespace voltdb {
...
 /**
  * Thread local key for storing integer value of amount of memory allocated
  */
-pthread_key_t m_allocatedKey;
-pthread_key_t m_threadPartitionIdKey;
-pthread_key_t m_enginePartitionIdKey;
-pthread_once_t m_keyOnce = PTHREAD_ONCE_INIT;
+    thread_local size_t* m_allocatedKey = nullptr;
+    thread_local int32_t* m_threadPartitionIdKey = nullptr;
+    thread_local int32_t* m_enginePartitionIdKey = nullptr;
```
Comment (JadenW) [inline src/ee/common/ThreadLocalPool.cpp]: We can drop the the "*Key" from the name of these variables, since they are no longer pthread keys.

File: src/ee/common/StringRef.cpp [hunk 10]
```diff
@@ -47,8 +49,15 @@ const char* StringRef::getObject(int32_t& lengthOut) const {
                              asSizedObject(m_stringPtr)->m_size)
               << std::endl;
     // */
-    lengthOut = asSizedObject(m_stringPtr)->m_size;
-    return asSizedObject(m_stringPtr)->m_data;
+    auto const* sized = asSizedObject(m_stringPtr);
```
[the remark this answers is not in the document]
Comment (kaikaizi) [reply src/ee/common/StringRef.cpp]: This is the place that 0-length string in ThreadLocalPool ends up being corrupted.
This force-casts some raw `char*` into `ThreadLocalPool::Sized` and retrieves the memory.
As to when a string/VARCHAR storage needs to be inlined, [this](https://wiki.voltdb.com/display/internal/Memory+Management+in+the+EE) has a better (probably still up to date) explanation.
For what I see, the empty string is not inlined.

File: src/ee/common/ThreadLocalPool.cpp (hunk 9 above)
Comment (kaikaizi) [reply src/ee/common/ThreadLocalPool.cpp]: Will do.

File: src/ee/common/ThreadLocalPool.h [hunk 11]
```diff
@@ -20,58 +20,39 @@
...
-    std::size_t* allocated;
-    int32_t* enginePartitionId;
+    PoolLocals(bool dummyEntry) { }
+    PoolLocals(const PoolLocals& src) = default;
+
+    PoolLocals& operator = (PoolLocals const& rhs) = default;
+
+    PoolPairType* poolData = nullptr;
+    CompactingStringStorage* stringData = nullptr;
+    std::size_t allocated;
```
Comment (n3world) [inline src/ee/common/ThreadLocalPool.h]: This is left uninitialized after the no arg constructor is called. Is that desired?

Comment (kaikaizi) [reply src/ee/common/ThreadLocalPool.h]: This is actually ok, and was intended that way. The ctor of PoolLocals initialize them from thread local variables (which is not well designed, to say at least), and `allocated` is assigned at that time. See ThreadLocalPool.cpp.
